source/common/router/header_parser.cc

```diff
@@ -286,12 +297,30 @@ void HeaderParser::evaluateHeaders(Http::HeaderMap& headers,
   for (const auto& [key, entry] : headers_to_add_) {
     const std::string value =
         stream_info != nullptr ? entry.formatter_->format(*stream_info) : entry.original_value_;
-    if (!value.empty()) {
-      if (entry.formatter_->append()) {
+    switch (entry.formatter_->appendAction()) {
+    case envoy::config::core::v3::HeaderValueOption::ADD_IF_ABSENT:
+      // Check whether the header already exist or not. We only
+      // need to add the header if it doesn't already exist.
+      if (headers.get(key).empty() && !value.empty()) {
```

Comment:
You check value.empty() at insertion time. Can you just assert here and below that it's not empty?

Reply:
Changed, Thanks!

Reply:
**UPDATE:** I realized that there are a few test cases which fails if I change this check to an assertion instead.

https://github.com/envoyproxy/envoy/blob/05ca792d25e4e40e472ba34ba4206fb69214929a/test/common/router/header_formatter_test.cc#L803

Reply:
Can you debug why they are failing? Is it because of a test issue? Or is it an issue that can actually happen?

/wait

Reply:
@mattklein123 
Looks like these three test cases actually assert on this behavior:
https://github.com/envoyproxy/envoy/blob/05ca792d25e4e40e472ba34ba4206fb69214929a/test/common/router/header_formatter_test.cc#L1089
https://github.com/envoyproxy/envoy/blob/05ca792d25e4e40e472ba34ba4206fb69214929a/test/common/router/header_formatter_test.cc#L1081
https://github.com/envoyproxy/envoy/blob/05ca792d25e4e40e472ba34ba4206fb69214929a/test/common/router/header_formatter_test.cc#L812

It doesn't make much sense to supply an empty value here, so I'm happy to fix these test cases and change the check to an asserttion as you suggested.

Reply:
Yes please just fix the tests if they don't make sense.

I will point out that adding a header with an empty value is OK. Are you sure we don't need to support this? It seems like maybe we should support this and the original code actually has a bug?

Reply:
How about we support it for `ADD_IF_ABSENT` and `OVERWRITE_IF_EXISTS` but not for `APPEND_IF_EXISTS`? It doesn't make sense to append an empty value to the existing header.

Reply:
I think that's fine as long as you have good test coverage for it all.

Reply:
Actually, I take what I said in 2 back. `APPEND_IF_EXISTS` also adds the header if it doesn't already exist. So I'll just drop the empty value check altogether and will add/modify tests accordingly. Thanks for your inputs!